wrapper/src/main/java/software/amazon/jdbc/plugin/endpoint/EndpointMonitor.java

```diff
@@ -0,0 +1,237 @@
+        TimeUnit.MILLISECONDS.sleep(this.intervalMs); // do not include this in the telemetry
+      } catch (final InterruptedException exception) {
+        LOGGER.finest(
+            () -> Messages.get(
+                "EndpointMonitor.interruptedExceptionDuringMonitoring",
+                new Object[] {this.hostSpec.getHost()}));
+      } catch (final Exception ex) {
+        // this should not be reached; log and exit thread
+        if (LOGGER.isLoggable(Level.FINEST)) {
+          LOGGER.log(
+              Level.FINEST,
+              Messages.get(
+                  "EndpointMonitor.exceptionDuringMonitoringStop",
+                  new Object[] {this.hostSpec.getHost()}),
+              ex); // We want to print full trace stack of the exception.
+        }
+      } finally {
+        if (telemetryContext != null) {
+          telemetryContext.closeContext();
+        }
+      }
+    }
+  }
+
+  private void openConnection() throws SQLException {
+    try {
+      if (this.monitoringConn == null || this.monitoringConn.isClosed()) {
+        // open a new connection
+        LOGGER.finest(() -> Messages.get(
+            "EndpointMonitor.openingConnection",
+            new Object[] {this.hostSpec.getUrl()}));
+        this.monitoringConn = this.pluginService.forceConnect(this.hostSpec, this.props);
+        LOGGER.finest(() -> Messages.get(
+            "EndpointMonitor.openedConnection",
+            new Object[] {this.monitoringConn}));
+      }
+    } catch (SQLException ex) {
+      if (this.monitoringConn != null && !this.monitoringConn.isClosed()) {
+        try {
+          this.monitoringConn.close();
+        } catch (Exception e) {
+          // ignore
+        }
+        this.monitoringConn = null;
+      }
+      throw ex;
+    }
+  }
+
+  private List<HostSpec> queryForEndpoints(final Connection conn) throws SQLException {
+    int networkTimeout = -1;
+    try {
+      networkTimeout = conn.getNetworkTimeout();
+      // The query is not monitored by the EFM plugin, so it needs a socket timeout
+      if (networkTimeout == 0) {
+        conn.setNetworkTimeout(networkTimeoutExecutor, defaultQueryTimeoutMs);
+      }
+    } catch (SQLException e) {
+      LOGGER.warning(() -> Messages.get("EndpointMonitor.getNetworkTimeoutError",
+          new Object[] {e.getMessage()}));
+    }
+
+    try (final Statement stmt = conn.createStatement();
+         final ResultSet resultSet = stmt.executeQuery("select * from test")) { // TODO: replace this mock table
+      return processQueryResults(resultSet);
+    } catch (final SQLSyntaxErrorException e) {
+      throw new SQLException(Messages.get("EndpointMonitor.invalidQuery"), e);
+    } finally {
+      if (networkTimeout == 0 && !conn.isClosed()) {
+        conn.setNetworkTimeout(networkTimeoutExecutor, networkTimeout);
+      }
+    }
+  }
+
+  private List<HostSpec> processQueryResults(final ResultSet resultSet) throws SQLException {
```

Comment:
Processing logic isn't too complex and it's just few lines of code. For readability, I'd suggest to move it to queryForEndpoints()

Reply:
Personally, I find this slightly easier to read. Do you feel strongly about this one? If so, I'll change it.